### PR TITLE
Implement #if, #else preprocessor for html files

### PR DESCRIPTION
**Pull request by eclecticdave**
This PR is my initial attempt to implement a preprocessor for generated html files (e.g. shell.html).  The goal is to replicate the existing JS preprocessor in python.

It isn't quite there yet, but I wanted to put it out there for comments.  #if , #else, #endif and #include are mostly implemented.  Things remaining to be done:

1. Token matching is a bit basic - it should probably use regexps.
2. There's no unary not, also the operator and value should be optional.
3. #define macros are not implemented yet.
4. I need to finish adding the test - it can be run manually for now.  (Also it should test #include as well).

### CODE REVIEW

**Comment by kripken** [comment]
Interesting, but what's the background/motivation here?

**Comment by eclecticdave** [comment]
@kripken:  The immediate use case is as discussed in PR #6080 - in order to potentially allow Module.print to switch behaviour dependent on the setting UNBUFFERED_PRINT.

However I also suspect it'll be useful when I resubmit my key event handler (representing the rest of PR #6003 which I'm splitting into two) which should only be included when using pthreads.  I also seem to recall @juj might have had a use for it as part of ASMFS, but I forget the details wrt that one.

**Comment by eclecticdave** [comment]
The browser test have mostly failed with 'no http activity' - which sounds like an issue with the test environment?

**Comment by eclecticdave** [comment]
Hi @kripken.  I've now completed my changes so this PR is ready for review.  It should be functionally the same as the JS preprocessor and I've added a test to the 'other' suite.

The test-browser failure on the bots seems to be unrelated,

I haven't implemented #define macros as these are not implemented in the JS preprocessor.  For some reason I thought they were, but I must have been imagining it :-)

**Comment by kripken** [comment]
I am concerned about duplicating code like this. Can the existing preprocessor not be used for this purpose?

In general, if we want to pick between having code in JS or in python, JS is probably more future-proof (as we may want to do more node.js integration going forward, e.g. webpack and other stuff). So adding a python duplicate, or replacing the JS one with python, doesn't seem like the right thing to me.

**Comment by kripken** [comment]
To be more concrete, we could make the existing preprocessor runnable by itself as `node src/preprocess.js`, by a little refactoring to move the code into there etc.

**Comment by eclecticdave** [comment]
Hmm, yes I considered that - but it doesn't seem to be entirely straightforward.  To get 'preprocessor.js' to run standalone we would have to add code to process the command line args, read the html file and call the preprocessor - but you don't want this code when you include it in compiler.js.

On the other hand if we put that extra code in a different file we then have to pull preprocessor.js in somehow.  We probably don't want to make it into a module if we're also loading it into compiler.js, so we would probably have to replicate the 'load' and 'read' functions from compiler.js and use those to pull in the function.  I don't know if it's OK to tie it to Node or if we want to keep it agnostic, in which case we're starting to need quite a few lines of additional code just to call a simple JS function from python.  Which at the time seemed like overkill to avoid a bit of duplication?

Or maybe there's an easier way of doing it and I've just lost the plot?  It wouldn't be the first time :-)

**Comment by kripken** [comment]
It should be ok to tie it to node.js, yeah.

Basically, I think we could have the core preprocessor in js, and a small python script could wrap around it, just doing any argument handling/conversion and doing the actual call to run the js preprocessor.

See also `tools/js-optimizer.js` which does `load('utility.js');` - that loads all the code in `src/utility.js`, which lets us share it between the js compiler and the js optimizer. Perhaps we can do something similar.

File: tools/shared.py [hunk 1]
```diff
@@ -2730,16 +2730,21 @@ def clang_preprocess(filename):
...
-  include_pattern = re.compile('^#include\s*["<](.*)[">]\s?$', re.MULTILINE)
-  while(1):
-    m = include_pattern.search(f, pos)
-    if not m:
-      return f
-    included_file = open(os.path.join(os.path.dirname(filename), m.groups(0)[0]), 'r').read()
+  temp_dir = configuration.get_temp_files().get_dir()
+
+  # Create a settings file with the current settings to pass to the JS preprocessor
+  settings_str = "var " + ";\nvar ".join(Settings.serialize()[1::2])
```
Comment (kripken) [inline tools/shared.py]: please explain in a comment  why we need `1::2` here. (I believe it is to skip the `-s` stuff from `-s K1=V1 -s K2=V2` etc.?)

File: tools/shared.py [hunk 2]
```diff
@@ -2730,16 +2730,21 @@ def clang_preprocess(filename):
...
+  settings_file = os.path.join(temp_dir, 'settings.js')
+  open(settings_file, 'w').write(settings_str)
 
-    f = f[:m.start(0)] + included_file + f[m.end(0):]
+  # Run the JS preprocessor
+  (path, file) = os.path.split(filename)
+  args=[settings_file, file]
```
Comment (kripken) [inline tools/shared.py]: space before and after `=`

File: tools/shared.py [hunk 3]
```diff
@@ -2730,16 +2730,21 @@ def clang_preprocess(filename):
...
+  stdout = os.path.join(temp_dir, 'stdout')
+  run_js(path_from_root('tools/preprocessor.js'), NODE_JS, args, True, stdout=open(stdout, 'w'), cwd=path)
```
Comment (kripken) [inline tools/shared.py]: do you need a temp file for stdout? i think you can use `PIPE` and just get the output from `run_js`?

Comment (kripken) [reply tools/shared.py]: (and in that case, i don't think `temp_dir` is needed, can just get a temp file for the settings file directly)

**Comment by eclecticdave** [comment]
@kripken - FYI I've made the requested changes and the CI failures here seem to be timeouts, rather than actual failures.

File: emcc.py [hunk 4]
```diff
@@ -2558,7 +2558,8 @@ def generate_html(target, options, js_target, target_basename,
   script = ScriptSource()
 
   logging.debug('generating HTML')
-  shell = open(options.shell_path).read()
+#  shell = open(options.shell_path).read()
```
Comment (kripken) [inline emcc.py]: unnecessary commented line

**Comment by kripken** [comment]
Same timeouts as before - which makes me think they might be real. Do you see them when running locally?

**Comment by eclecticdave** [comment]
@kripken - Looks like I need to investigate this - anything that goes through the preprocessor now seems to be taking much longer than it should.

More worryingly there seem to be some failures in circleci: test-other, even though it says it passed.  Not sure why the test harness failed to spot that.

**Comment by eclecticdave** [comment]
@kripken - I think I've found the problem.  I've reverted the change where we stopped run_js from writing stdout to a file.
Apparently there's a long standing python bug with Popen and stdout=PIPE which results in the subprocess hanging if it outputs more than 64k - and shell.html is larger than that!

**Comment by kripken** [comment]
Good find! Now that you mention it, it does sound familiar - I think we may have hit it before in this project some years ago... I could be mixing it up though.

**Comment by kripken** [comment]
Oops, I think we did the wrong thing here: After you noticed test errors were being ignored, we fixed it on incoming - but incoming would have had to be merged into here to pick that up. So we merged this with errors.

I pushed a one-line change to revert the line that actually calls the preprocessor. A new PR can re-enable it with those tests fixed.

Log with list of tests that broke: https://circleci.com/gh/kripken/emscripten/23617?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by kripken** [comment]
I also had to disable `other.test_html_preprocess`, we should re-enable it when we re-enable this.

**Comment by eclecticdave** [comment]
Sorry about that :-( ... I'll take a look at those test failures tonight and prepare another PR

**Pull request merged**